src/iterable_jsonb.rs

```diff
@@ -0,0 +1,253 @@
+                })
+            })
+        }
+    }
+}
+
+impl Drop for Jsonb<'_> {
+    fn drop(&mut self) {
+        if self.needs_drop {
+            unsafe {
+                pfree(self.pg_jsonb as void_mut_ptr);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Token<'a> {
+    BeginArray,
+    EndArray,
+    BeginObject,
+    EndObject,
+    /// Reresents a key in a JSON object.
+    /// Always followed a non-key [`Token`],
+    /// representing an associated value.
+    Key(&'a str),
+    // Scalar values used as array elements and object values
+    Null,
+    Bool(bool),
+    String(&'a str),
+    Numeric(JsonbNormalizedNumeric), // Normalized with numeric_normalize
+}
+
+pub struct TokenIterator<'a> {
+    pg_jsonb_iter: *mut JsonbIterator,
+    raw_scalar: bool,
+    _marker: std::marker::PhantomData<Jsonb<'a>>,
+}
+
+impl<'a> Iterator for TokenIterator<'a> {
+    type Item = Token<'a>;
+
+    #[allow(non_upper_case_globals)]
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut jsonb_val = JsonbValue::default();
+        let r: JsonbIteratorToken = unsafe {
+            JsonbIteratorNext(
+                &mut self.pg_jsonb_iter as *mut *mut JsonbIterator,
+                &mut jsonb_val as *mut JsonbValue,
+                false,
+            )
+        };
+        match r {
+            JsonbIteratorToken_WJB_DONE => None,
+            JsonbIteratorToken_WJB_BEGIN_ARRAY => {
+                if unsafe { jsonb_val.val.array.rawScalar } {
+                    self.raw_scalar = true;
+                    self.next()
+                } else {
+                    Some(Token::BeginArray)
+                }
+            }
+            JsonbIteratorToken_WJB_END_ARRAY => {
+                if self.raw_scalar {
+                    self.next()
+                } else {
+                    Some(Token::EndArray)
+                }
+            }
+            JsonbIteratorToken_WJB_BEGIN_OBJECT => Some(Token::BeginObject),
+            JsonbIteratorToken_WJB_END_OBJECT => Some(Token::EndObject),
+            JsonbIteratorToken_WJB_KEY => {
+                match TokenIterator::extract_value_token(&mut jsonb_val) {
+                    Token::String(str) => Some(Token::Key(str)),
+                    _ => {
+                        elog(PgLogLevel::ERROR, "Unexpected token while expecting a key");
+                        None
+                    }
+                }
+            }
+            JsonbIteratorToken_WJB_VALUE | JsonbIteratorToken_WJB_ELEM => {
+                Some(TokenIterator::extract_value_token(&mut jsonb_val))
+            }
+            _ => {
+                elog(
+                    PgLogLevel::ERROR,
+                    format!("invalid JsonbIteratorNext rc: {}", r).as_str(),
+                );
+                None
+            }
+        }
+    }
+}
+
+impl<'a> TokenIterator<'a> {
+    #[inline]
+    fn extract_string_value(jsonb_val: &mut JsonbValue) -> &'a str {
```

Comment:
nit: Here's a potential candidate for `unsafe fn`. It's only safe to call this when the caller has validated the `JsonbValue`'s type is 'jbvString'.